doc/topics/writing-recipes.rst

```diff
@@ -0,0 +1,355 @@
+  return their paths.  Doing so allows Buildout to take care of
+  removing them if the part is uninstalled, making a separate
+  uninstall recipe unnecessary.
+
+To make these responsibilities easier to cope with, the ``option``
+object passed to the constructor has a helper function, ``created``.
+It should be passed one or more paths just before they are created and
+returns a list of all of the paths passed as well as any earlier paths
+created.  If an exception is raised, any files or directories created
+will be removed automatically. When the recipe returns, it can just
+return the result of calling ``created()`` with no arguments.
+
+Example: configuration from template recipe
+-------------------------------------------
+
+In this example, we'll show a recipe that creates a configuration file
+based on a configuration string computed using value substitutions
+[#templaterecipes]_.  A sample usage:
+
+.. code-block:: ini
+
+   [buildout]
+   develop = src
+   parts = server
+
+   [config]
+   recipe = democonfigrecipe
+   port = 8080
+   contents =
+      <zeo>
+        address ${:port}
+      </zeo>
+      <mappingstorage>
+      </mappingstorage>
+
+   [server]
+   recipe = zc.zdaemonrecipe
+   program = runzeo -C ${config:path}
+
+.. -> src
+
+   >>> write(src, 'buildout.cfg')
+
+Some things to note about this example:
+
+- The ``config`` part uses the recipe whose source code we'll show
+  below.  It has a ``port`` option, which it uses in its ``contents``
+  option.  It could as easily have used options from other sections.
+
+- The ``server`` part uses ``${config:path}`` to get the path to the
+  configuration file generated by the ``config`` part. The ``path``
+  option value will be computed by the recipe for use in other parts,
+  as we've seen here.
+
+- We didn't have to list the ``config`` part in the buildout ``parts``
+  option.  It's :doc:`added implicitly <implicit-parts>` by virtue of
+  its use in the ``server`` part.
+
+- We used the ``develop`` option to specify a ``src`` directory
+  containing our recipe.  This allows us to use the recipe locally
+  without having to build a distribution file.
+
+If we were to run this buildout, a ``parts/config`` file would be
+generated:
+
+.. code-block:: xml
+
+   <zeo>
+     address 8080
+   </zeo>
+   <mappingstorage>
+   </mappingstorage>
+
+.. -> config
+
+as would a zdaemon configuration file, ``parts/server/zdaemon.conf``, like:
+
+.. code-block:: xml
+
+    <runner>
+      daemon on
+      directory /sample/parts/server
+      program runzeo -C /sample/parts/config
+      socket-name /sample/parts/server/zdaemon.sock
+      transcript /sample/parts/server/transcript.log
+    </runner>
+
+    <eventlog>
+      <logfile>
+        path /sample/parts/server/transcript.log
+      </logfile>
+    </eventlog>
+
+.. -> server
+
+   >>> server = server.replace('/sample', os.getcwd())
+
+Here's the recipe source, ``src/democonfigrecipe.py``::
```

Comment:
You probably want that fancy `.. code-block:: python` directive here, to get syntax highlighting.